### PR TITLE
Admin users

### PR DESCRIPTION
**Pull request by AlonaDorozhy**
### Minor changes in the component.
**At the moment, all Get, Put, and Delete are working correctly.**

### CODE REVIEW

File: src/app/admin-users/helpers/helper-details/helper-details.component.ts [hunk 1]
```diff
@@ -19,16 +20,17 @@ export class HelperDetailsComponent implements OnInit, OnDestroy {
   public profileForm: FormGroup;
   public isEditDisabled: boolean;
   public visibleFields = false;
+  public permissions = Permissions;
   private unsubscribe: Subject<void> = new Subject();
-
+  private changedPhoto: string | ArrayBuffer;
```
Comment (VoropN) [inline src/app/admin-users/helpers/helper-details/helper-details.component.ts]: this property has to be public

File: src/app/admin-users/helpers/helper-details/helper-details.component.ts [hunk 2]
```diff
@@ -98,4 +100,18 @@ export class HelperDetailsComponent implements OnInit, OnDestroy {
     this.isEditDisabled = this.profileForm.disabled;
   }
 
+  private changeAvatar(event) {
+    const fileReader = new FileReader();
+    if (event && event.length) {
+      fileReader.readAsDataURL(event && event.length && event[0]);
```
Comment (VoropN) [inline src/app/admin-users/helpers/helper-details/helper-details.component.ts]: use here only event[0]

File: src/app/admin-users/managers/manager-details/manager-details.component.ts [hunk 3]
```diff
@@ -18,16 +19,19 @@ export class ManagerDetailsComponent implements OnInit, OnDestroy {
   public managerId: string;
   public profileForm: FormGroup;
   public isEditDisabled: boolean;
+  public permissions = Permissions;
   public visibleFields = false;
+  private changedPhoto: string | ArrayBuffer;
```
Comment (VoropN) [inline src/app/admin-users/managers/manager-details/manager-details.component.ts]: this property has to be public

File: src/app/admin-users/managers/manager-details/manager-details.component.ts [hunk 4]
```diff
@@ -97,4 +101,18 @@ export class ManagerDetailsComponent implements OnInit, OnDestroy {
     this.isEditDisabled = this.profileForm.disabled;
   }
 
+  private changeAvatar(event) {
```
Comment (VoropN) [inline src/app/admin-users/managers/manager-details/manager-details.component.ts]: this method has to be public

File: src/app/admin-users/helpers/services/helper.service.ts [hunk 5]
```diff
@@ -25,24 +25,24 @@ export class HelpersService {
...
       concatMap((config: Config) =>
-        this.http.get<Helper>(`${config.helpersApi}${id}`)
+        this.http.get<Helper>(`${config.helpersApi}/${id}`)
       )
     );
   }
 
 
-  public updateHelperById(formValue: object, id: string): Observable<Helper> {
+  public updateHelperById(formValue: object, id: string, avatar): Observable<Helper> {
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/services/helper.service.ts]: do not use object as a type, create a proper model for it, the same for avatar - add type if you need it

File: src/app/admin-users/managers/services/manager.service.ts [hunk 6]
```diff
@@ -27,23 +27,24 @@ export class ManagersService {
...
         return this.configService.getConfig().pipe(
             concatMap((config: Config) =>
-                this.http.get<Manager>(`${config.managersApi}${id}`)
+                this.http.get<Manager>(`${config.managersApi}/${id}`)
             )
         );
     }
 
-    public updateManagerById(formValue: object, id: string): Observable<Manager> {
+    public updateManagerById(formValue: object, id: string, avatar): Observable<Manager> {
```
Comment (ksenya-ta) [inline src/app/admin-users/managers/services/manager.service.ts]: same here, add proper types

File: src/app/admin-users/managers/services/manager.service.ts [hunk 7]
```diff
@@ -27,23 +27,24 @@ export class ManagersService {
...
+        console.log(JSON.stringify(formValue));
```
Comment (ksenya-ta) [inline src/app/admin-users/managers/services/manager.service.ts]: remove console.log

File: src/app/admin-users/styles/_list.scss [hunk 8]
```diff
@@ -4,4 +4,5 @@
 
 button{
     margin: 30px 15px 0 15px;
+
```
Comment (ksenya-ta) [inline src/app/admin-users/styles/_list.scss]: remove extra empty line

File: src/app/admin-users/managers/manager-details/manager-details.component.html [hunk 9]
```diff
@@ -1,19 +1,19 @@
...
-        </div>
         <mat-card-content>
             <form class="details-form" [formGroup]="profileForm">
-                <mat-form-field *ngIf="visibleFields" class="details-full-width">
-                    <input matInput placeholder="Path to photo" formControlName="photoPath">
-                </mat-form-field>
+                <div *ngIf="manager" class="heder-card">
+                    <h2> All Information about {{ manager.name }}</h2>
+                    <img mat-card-image
+                    [src]="changedPhoto ? changedPhoto : manager.photoPath"
```
Comment (ksenya-ta) [inline src/app/admin-users/managers/manager-details/manager-details.component.html]: is it photoPath or avatar? (it has here photoPath and avatar in other code)

File: src/app/admin-users/managers/manager-details/manager-details.component.ts [hunk 10]
```diff
@@ -18,16 +19,19 @@ export class ManagerDetailsComponent implements OnInit, OnDestroy {
   public managerId: string;
   public profileForm: FormGroup;
   public isEditDisabled: boolean;
+  public permissions = Permissions;
```
Comment (ksenya-ta) [inline src/app/admin-users/managers/manager-details/manager-details.component.ts]: do you use it?

File: src/app/admin-users/helpers/helper-details/helper-details.component.ts [hunk 11]
```diff
@@ -19,16 +20,17 @@ export class HelperDetailsComponent implements OnInit, OnDestroy {
   public profileForm: FormGroup;
   public isEditDisabled: boolean;
   public visibleFields = false;
+  public permissions = Permissions;
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/helper-details/helper-details.component.ts]: do you use it?

File: src/app/admin-users/admin-users.module.ts [hunk 12]
```diff
@@ -18,6 +18,8 @@ import { MatTabsModule } from '@angular/material/tabs';
 import { MatInputModule } from '@angular/material/input';
 import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 import { SearchModule } from '../search/search.module';
+import {MatDialogModule} from '@angular/material/dialog';
```
Comment (ksenya-ta) [inline src/app/admin-users/admin-users.module.ts]: add missing spaces

**Comment by ksenya-ta** [comment]
what the difference between helpers and managers? why do we need 2 services/components if all the code is almost identical for both this entities?

File: src/app/admin-users/helpers/helper-details/helper-details.component.html [hunk 13]
```diff
@@ -1,19 +1,18 @@
...
-            <mat-card-content>
-                <p><img mat-card-image class="avatar" [src]="helper.photoPath" [alt]="helper.name"></p>
-            </mat-card-content>
-        </div>
         <mat-card-content>
-            <form class="details-form" [formGroup]="profileForm">
-                <mat-form-field *ngIf="visibleFields" class="details-full-width">
-                    <input matInput placeholder="Path to photo" formControlName="photoPath">
-                </mat-form-field>
+            <form #f="ngForm" class="details-form" [formGroup]="profileForm">
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/helper-details/helper-details.component.html]: why do you need this #f?

File: src/app/admin-users/helpers/helper-details/helper-details.component.ts [hunk 14]
```diff
@@ -19,16 +19,17 @@ export class HelperDetailsComponent implements OnInit, OnDestroy {
...
   public visibleFields = false;
+  public changedPhoto: string | ArrayBuffer;
   private unsubscribe: Subject<void> = new Subject();
 
   constructor(
     private helpersService: HelpersService,
     private activatedRoute: ActivatedRoute,
     private fb: FormBuilder,
-    private router: Router
+    private router: Router,
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/helper-details/helper-details.component.ts]: do not leave trailing commas, especially in the method definitions

File: src/app/admin-users/helpers/helper-details/helper-details.component.ts [hunk 15]
```diff
@@ -57,37 +58,69 @@ export class HelperDetailsComponent implements OnInit, OnDestroy {
...
+      name: this.profileForm.get('name').value,
+      surname: this.profileForm.get('surname').value,
+      patronymic: this.profileForm.get('patronymic').value,
+      birthday: this.profileForm.get('birthday').value,
+      rating: this.profileForm.get('rating').value,
+      emailID: this.helper.emailID,
+      avatar: this.helper.avatar,
+      addressID: this.helper.addressID,
+      photoPath: this.helper.photoPath,
+    };
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/helper-details/helper-details.component.ts]: why don't you keep all this info in the form so you can just use all the form value for update?

File: src/app/admin-users/helpers/helper-details/helper-details.component.ts [hunk 16]
```diff
@@ -57,37 +58,69 @@ export class HelperDetailsComponent implements OnInit, OnDestroy {
...
     this.profileForm.patchValue({
+      id: helper.id,
       name: helper.name,
       surname: helper.surname,
       patronymic: helper.patronymic,
       birthday: helper.birthday,
+      rating: helper.rating,
+      emailID: helper.emailID,
       avatar: helper.avatar,
       photoPath: helper.photoPath,
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/helper-details/helper-details.component.ts]: do not leave trailing commas

File: src/app/admin-users/helpers/services/helper.service.ts [hunk 17]
```diff
@@ -10,41 +10,60 @@ import { Config } from 'src/app/shared/services/config/config.interface';
...
+  };
 
   public getAllHelpers(paramObj: object = {}): Observable<Helper[]> {
     let params = new HttpParams();
     Object.keys(paramObj).forEach((key: string) => params = params.set(key, paramObj[key]));
     return this.configService.getConfig().pipe(
       concatMap((config: Config) =>
         this.http.get<Helper[]>(config.helpersApi, { params })
-      )
+      ),
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/services/helper.service.ts]: why this comma is here?

File: src/app/admin-users/helpers/services/helper.service.ts [hunk 18]
```diff
@@ -10,41 +10,60 @@ import { Config } from 'src/app/shared/services/config/config.interface';
...
 
-  public updateHelperById(formValue: object, id: string): Observable<Helper> {
+  public updateHelperById(helperData: Helper): Observable<Helper> {
     return this.configService.getConfig().pipe(
-      concatMap((config: Config) =>
-        this.http.put<Helper>(`${config.helpersApi}${id}`, formValue)
-      )
+      concatMap((config: Config) => {
+        return this.http.put<Helper[]>(`${config.helpersApi}/${helperData.id}`, helperData, this.httpOptions),
+          this.putHelperImage(config.managersImageApi, helperData, helperData.id);
```
Comment (ksenya-ta) [inline src/app/admin-users/helpers/services/helper.service.ts]: please explain this code, what is the expected behavior here? why these 2 api calls added one by one using comma?

File: src/app/admin-users/managers/managers-list/managers-list.component.ts [hunk 19]
```diff
@@ -12,9 +13,9 @@ import { ManagersService } from '../services/manager.service';
 export class ManagersListComponent implements OnInit {
   public managers$: Observable<Manager[]>;
 
-  constructor(private managerService: ManagersService) { }
+  constructor(private managerService: ManagersService, public dialog: MatDialog) { }
```
Comment (ksenya-ta) [inline src/app/admin-users/managers/managers-list/managers-list.component.ts]: do you use this added dialog?

File: src/app/admin-users/managers/services/manager.service.ts [hunk 20]
```diff
@@ -27,25 +31,38 @@ export class ManagersService {
...
+    }
+
+    private createFormData(info: Manager): FormData {
+        const formData = new FormData();
+        Object.entries(info).forEach(([key, value]: [string, Blob]) =>
+            value instanceof File ?
+                formData.append(key, value, value.name) :
+                formData.append(key, value));
+        return formData;
+    }
```
Comment (ksenya-ta) [inline src/app/admin-users/managers/services/manager.service.ts]: do not copy paste code. if you need it twice - reuse it

File: src/app/admin-users/styles/_item.scss [hunk 21]
```diff
@@ -1,20 +1,26 @@
...
 .avatar {
-    width: 100px;
+    max-width: 130px;
+    width: auto;
+    max-height: 130px;
     float: left;
-    margin: 10px;
+    margin: 0 10px 0 10px;
+}
+.content{
```
Comment (ksenya-ta) [inline src/app/admin-users/styles/_item.scss]: fix formatting

File: src/app/navbar/navbar.component.ts [hunk 22]
```diff
@@ -1,10 +1,12 @@
 import { Component } from '@angular/core';
+import {Permissions} from 'src/app/shared/permissions/models/permissions.enum';
```
Comment (ksenya-ta) [inline src/app/navbar/navbar.component.ts]: spaces are missing

File: src/app/admin-users/services/admin-user.service.ts [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+    return this.updateUserById<Manager>(helperData, { id: helperData.id, api: 'managersApi', apiForImage: 'managersImageApi' });
+  }
+
+  private updateUserById<T>(data: T, argObj: EditUserModel) {
+    this.configService.getConfig().pipe(
+      concatMap((config: Config) => {
+        return this.http.put<T>(`${config[argObj.api]}/${argObj.id}`, data, this.httpOptions);
+      })
+    );
+    return this.putImage<T>(data, argObj);
```
Comment (ksenya-ta) [inline src/app/admin-users/services/admin-user.service.ts]: no comma in here anymore, it is good. however it still will perform only second request (that saves image). Are you sure it is expected behavior?

File: src/app/admin-users/managers/managers-list/managers-list.component.spec.ts [hunk 24]
```diff
@@ -1,25 +1,79 @@
...
+    managerService = fixture.debugElement.injector.get(AdminUserService);
   });
 
-  xit('should create', () => {
-    expect(component).toBeTruthy();
+  it('Expect to receive list of managers', () => {
+    spyOn(managerService, 'getAllManagers').and.returnValue(of(mockManagers));
+    component.ngOnInit();
+    component.managers$.subscribe((managers: Manager[]) => {
+      expect(managers).toEqual(mockManagers);
```
[the remark this answers is not in the document]
Comment (AlonaDorozhy) [reply src/app/admin-users/managers/managers-list/managers-list.component.spec.ts]: I added done, for some reason it is not displayed here

File: src/app/admin-users/helpers/helpers-list/helpers-list.component.spec.ts [hunk 25]
```diff
@@ -1,25 +1,78 @@
...
+    helperService = fixture.debugElement.injector.get(AdminUserService);
   });
 
-  xit('should create', () => {
-    expect(component).toBeTruthy();
+  it('Expect to receive list of helpers', () => {
+    spyOn(helperService, 'getAllHelpers').and.returnValue(of(mockHelpers));
+    component.ngOnInit();
+    component.helpers$.subscribe((helpers: Helper[]) => {
+      expect(helpers).toEqual(mockHelpers);
```
[the remark this answers is not in the document]
Comment (AlonaDorozhy) [reply src/app/admin-users/helpers/helpers-list/helpers-list.component.spec.ts]: I added done already

**Pull request merged**